File: ios/Classes/NpuPoseDetectionPlugin.swift
import Flutter
import UIKit
import Vision
import CoreMedia
import CoreVideo

/// Flutter plugin for NPU-accelerated pose detection on iOS.
///
/// Uses Apple Vision Framework with automatic Neural Engine acceleration.
public class NpuPoseDetectionPlugin: NSObject, FlutterPlugin {

    // MARK: - Properties

    private var poseDetector: VisionPoseDetector?
    private var config: DetectorConfig = DetectorConfig()
    private var frameEventSink: FlutterEventSink?
    private var videoProgressEventSink: FlutterEventSink?
    private var isStreamingActive = false
    private var frameCount: Int = 0
    private var fpsCalculator = FPSCalculator()
    private var videoProcessor: VideoProcessor?

    // MARK: - Channel Names

    private static let methodChannelName = "com.example.npu_pose_detection/methods"
    private static let eventChannelName = "com.example.npu_pose_detection/frames"
    private static let videoProgressChannelName = "com.example.npu_pose_detection/video_progress"

    // MARK: - Plugin Registration

    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(
            name: methodChannelName,
            binaryMessenger: registrar.messenger()
        )
        let instance = NpuPoseDetectionPlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        // Setup EventChannel for camera frame streaming
        let eventChannel = FlutterEventChannel(
            name: eventChannelName,
            binaryMessenger: registrar.messenger()
        )
        eventChannel.setStreamHandler(instance)

        // Setup EventChannel for video analysis progress
        let videoProgressChannel = FlutterEventChannel(
            name: videoProgressChannelName,
            binaryMessenger: registrar.messenger()
        )
        videoProgressChannel.setStreamHandler(VideoProgressStreamHandler(plugin: instance))
    }

    // MARK: - Method Call Handler

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            handleInitialize(call, result: result)
        case "detectPose":
            handleDetectPose(call, result: result)
        case "detectPoseFromFile":
            handleDetectPoseFromFile(call, result: result)
        case "processFrame":
            handleProcessFrame(call, result: result)
        case "startCameraDetection":
            handleStartCameraDetection(call, result: result)
        case "stopCameraDetection":
            handleStopCameraDetection(result: result)
        case "analyzeVideo":
            handleAnalyzeVideo(call, result: result)
        case "cancelVideoAnalysis":
            handleCancelVideoAnalysis(result: result)
        case "updateConfig":
            handleUpdateConfig(call, result: result)
        case "getDeviceCapabilities":
            handleGetDeviceCapabilities(result: result)
        case "dispose":
            handleDispose(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Initialize

    private func handleInitialize(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let configDict = args["config"] as? [String: Any] else {
            result(errorResponse(code: "invalidArguments", message: "Missing config"))
            return
        }

        config = DetectorConfig.from(dictionary: configDict)
        poseDetector = VisionPoseDetector(config: config)

        do {
            let mode = try poseDetector?.initialize()
            result([
                "success": true,
                "accelerationMode": mode?.rawValue ?? "npu",
                "modelVersion": "vision_body_pose_v1"
            ])
        } catch {
            result(errorResponse(
                code: "modelLoadFailed",
                message: "Failed to initialize Vision detector",
                platformMessage: error.localizedDescription
            ))
        }
    }

    // MARK: - Detect Pose

    private func handleDetectPose(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let detector = poseDetector else {
            result(errorResponse(code: "notInitialized", message: "Detector not initialized"))
            return
        }

        guard let args = call.arguments as? [String: Any],
              let imageDataBase64 = args["imageData"] as? String,
              let imageData = Data(base64Encoded: imageDataBase64) else {
            result(errorResponse(code: "invalidImageFormat", message: "Invalid image data"))
            return
        }

        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let poseResult = try detector.detectPose(imageData: imageData)
                DispatchQueue.main.async {
                    result([
                        "success": true,
                        "result": poseResult.toDictionary()
                    ])
                }
            } catch {
                DispatchQueue.main.async {
                    result(self.errorResponse(
                        code: "inferenceFailed",
                        message: "Pose detection failed",
                        platformMessage: error.localizedDescription
                    ))
                }
            }
        }
    }

    // MARK: - Detect Pose From File

    private func handleDetectPoseFromFile(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let detector = poseDetector else {
            result(errorResponse(code: "notInitialized", message: "Detector not initialized"))
            return
        }

        guard let args = call.arguments as? [String: Any],
              let filePath = args["filePath"] as? String else {
            result(errorResponse(code: "invalidArguments", message: "Missing file path"))
            return
        }

        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let poseResult = try detector.detectPose(filePath: filePath)
                DispatchQueue.main.async {
                    result([
                        "success": true,
                        "result": poseResult.toDictionary()
                    ])
                }
            } catch {
                DispatchQueue.main.async {
                    result(self.errorResponse(
                        code: "inferenceFailed",
                        message: "Pose detection failed",
                        platformMessage: error.localizedDescription
                    ))
                }
            }
        }
    }

    // MARK: - Update Config

    private func handleUpdateConfig(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let configDict = args["config"] as? [String: Any] else {
            result(errorResponse(code: "invalidArguments", message: "Missing config"))
            return
        }

        config = DetectorConfig.from(dictionary: configDict)
        poseDetector?.updateConfig(config)

        result([
            "success": true,
            "requiresReinitialize": false
        ])
    }

    // MARK: - Device Capabilities

    private func handleGetDeviceCapabilities(result: @escaping FlutterResult) {
        let device = UIDevice.current
        let hasNeuralEngine = hasNeuralEngineSupport()

        result([
            "success": true,
            "capabilities": [
                "platform": "ios",
                "osVersion": device.systemVersion,
                "deviceModel": getDeviceModel(),
                "npuInfo": [
                    "type": "neural_engine",
                    "isAvailable": hasNeuralEngine,
                    "performanceTier": hasNeuralEngine ? "high" : "low"
                ],
                "hasGpuSupport": true,
                "recommendedMode": "npu"
            ]
        ])
    }

    // MARK: - Process Frame

    private func handleProcessFrame(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let detector = poseDetector else {
            result(errorResponse(code: "notInitialized", message: "Detector not initialized"))
            return
        }

        guard let args = call.arguments as? [String: Any],
              let planes = args["planes"] as? [[String: Any]],
              let width = args["width"] as? Int,
              let height = args["height"] as? Int else {
            result(errorResponse(code: "invalidArguments", message: "Missing frame data"))
            return
        }

        DispatchQueue.global(qos: .userInitiated).async {
            do {
                // Create pixel buffer from plane data
                guard let pixelBuffer = self.createPixelBuffer(
                    from: planes,
                    width: width,
                    height: height
                ) else {
                    DispatchQueue.main.async {
                        result(self.errorResponse(code: "invalidImageFormat", message: "Failed to create pixel buffer"))
                    }
                    return
                }

                let poseResult = try detector.processFrame(pixelBuffer: pixelBuffer)
                DispatchQueue.main.async {
                    result([
                        "success": true,
                        "result": poseResult.toDictionary()
                    ])
                }
            } catch {
                DispatchQueue.main.async {
                    result(self.errorResponse(
                        code: "inferenceFailed",
                        message: "Frame processing failed",
                        platformMessage: error.localizedDescription
                    ))
                }
            }
        }
    }

    // MARK: - Camera Detection

    private func handleStartCameraDetection(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard poseDetector != nil else {
            result(errorResponse(code: "notInitialized", message: "Detector not initialized"))
            return
        }

        isStreamingActive = true
        frameCount = 0
        fpsCalculator.reset()

        result(["success": true])
    }

    private func handleStopCameraDetection(result: @escaping FlutterResult) {
        isStreamingActive = false

        // Send end event
        frameEventSink?([
            "type": "end",
            "reason": "stopped"
        ])

        result(["success": true])
    }

    /// Process a frame and send result through event channel.
    /// Called from Dart side when a camera frame is available.
    func processFrameForStream(planes: [[String: Any]], width: Int, height: Int, timestampUs: Int) {
        guard isStreamingActive, let detector = poseDetector else { return }

        DispatchQueue.global(qos: .userInitiated).async {
            do {
                guard let pixelBuffer = self.createPixelBuffer(
                    from: planes,
                    width: width,
                    height: height
                ) else { return }

                let poseResult = try detector.processFrame(pixelBuffer: pixelBuffer)

                self.frameCount += 1
                let fps = self.fpsCalculator.update()

                DispatchQueue.main.async {
                    self.frameEventSink?([
                        "type": "frame",
                        "frameNumber": self.frameCount,
                        "timestampUs": timestampUs,
                        "fps": fps,
                        "result": poseResult.toDictionary()
                    ])
                }
            } catch {
                DispatchQueue.main.async {
                    self.frameEventSink?([
                        "type": "error",
                        "error": [
                            "code": "inferenceFailed",
                            "message": "Frame processing failed",
                            "platformMessage": error.localizedDescription
                        ]
                    ])
                }
            }
        }
    }

    // MARK: - Pixel Buffer Creation

    private func createPixelBuffer(from planes: [[String: Any]], width: Int, height: Int) -> CVPixelBuffer? {
        guard let firstPlane = planes.first,
              let bytesBase64 = firstPlane["bytes"] as? String,
              let bytes = Data(base64Encoded: bytesBase64) else {
            return nil
        }

        var pixelBuffer: CVPixelBuffer?
        let status = CVPixelBufferCreate(
            kCFAllocatorDefault,
            width,
            height,
            kCVPixelFormatType_32BGRA,
            nil,
            &pixelBuffer
        )

        guard status == kCVReturnSuccess, let buffer = pixelBuffer else {
            return nil
        }

        CVPixelBufferLockBaseAddress(buffer, [])
        defer { CVPixelBufferUnlockBaseAddress(buffer, []) }

        if let baseAddress = CVPixelBufferGetBaseAddress(buffer) {
            bytes.copyBytes(to: baseAddress.assumingMemoryBound(to: UInt8.self), count: bytes.count)
        }

        return buffer
    }

    // MARK: - Video Analysis

    private func handleAnalyzeVideo(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let detector = poseDetector else {
            result(errorResponse(code: "notInitialized", message: "Detector not initialized"))
            return
        }

        guard let args = call.arguments as? [String: Any],
              let videoPath = args["videoPath"] as? String else {
            result(errorResponse(code: "invalidArguments", message: "Missing video path"))
            return
        }

        let frameInterval = args["frameInterval"] as? Int ?? 1

        videoProcessor = VideoProcessor(detector: detector)

        videoProcessor?.analyzeVideo(
            at: videoPath,
            frameInterval: frameInterval,
            progress: { [weak self] progress in
                self?.videoProgressEventSink?(progress.toDictionary())
            },
            completion: { [weak self] analysisResult in
                switch analysisResult {
                case .success(let videoResult):
                    self?.videoProgressEventSink?([
                        "type": "complete"
                    ])
                    result([
                        "success": true,
                        "result": videoResult.toDictionary()
                    ])
                case .failure(let error):
                    self?.videoProgressEventSink?([
                        "type": "error",
                        "error": [
                            "code": "videoAnalysisFailed",
                            "message": error.localizedDescription
                        ]
                    ])
                    result(self?.errorResponse(
                        code: "videoAnalysisFailed",
                        message: "Video analysis failed",
                        platformMessage: error.localizedDescription
                    ))
                }
            }
        )
    }

    private func handleCancelVideoAnalysis(result: @escaping FlutterResult) {
        videoProcessor?.cancel()
        videoProgressEventSink?([
            "type": "cancelled"
        ])
        result(["success": true])
    }

    // MARK: - Dispose

    private func handleDispose(result: @escaping FlutterResult) {
        isStreamingActive = false
        videoProcessor?.cancel()
        videoProcessor = nil
        poseDetector?.dispose()
        poseDetector = nil
        result(["success": true])
    }

    // MARK: - Helpers

    private func errorResponse(
        code: String,
        message: String,
        platformMessage: String? = nil,
        recoverySuggestion: String? = nil
    ) -> [String: Any] {
        var error: [String: Any] = [
            "code": code,
            "message": message
        ]
        if let pm = platformMessage {
            error["platformMessage"] = pm
        }
        if let rs = recoverySuggestion {
            error["recoverySuggestion"] = rs
        }
        return [
            "success": false,
            "error": error
        ]
    }

    private func hasNeuralEngineSupport() -> Bool {
        // Neural Engine available on A12+ chips (iPhone XS and later)
        if #available(iOS 14.0, *) {
            return true
        }
        return false
    }

    private func getDeviceModel() -> String {
        var systemInfo = utsname()
        uname(&systemInfo)
        let machineMirror = Mirror(reflecting: systemInfo.machine)
        let identifier = machineMirror.children.reduce("") { identifier, element in
            guard let value = element.value as? Int8, value != 0 else { return identifier }
            return identifier + String(UnicodeScalar(UInt8(value)))
        }
        return identifier
    }
}

// MARK: - FlutterStreamHandler

extension NpuPoseDetectionPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        frameEventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        frameEventSink = nil
        isStreamingActive = false
        return nil
    }
}

// MARK: - Video Progress Stream Handler

class VideoProgressStreamHandler: NSObject, FlutterStreamHandler {
    weak var plugin: NpuPoseDetectionPlugin?

    init(plugin: NpuPoseDetectionPlugin) {
        self.plugin = plugin
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        plugin?.setVideoProgressEventSink(events)
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        plugin?.setVideoProgressEventSink(nil)
        return nil
    }
}

extension NpuPoseDetectionPlugin {
    func setVideoProgressEventSink(_ sink: FlutterEventSink?) {
        videoProgressEventSink = sink
    }
}

// MARK: - FPS Calculator

class FPSCalculator {
    private var timestamps: [CFAbsoluteTime] = []
    private let windowSize = 30

    func reset() {
        timestamps.removeAll()
    }

    func update() -> Double {
        let now = CFAbsoluteTimeGetCurrent()
        timestamps.append(now)

        // Keep only recent timestamps
        if timestamps.count > windowSize {
            timestamps.removeFirst()
        }

        // Calculate FPS from time window
        guard timestamps.count >= 2 else { return 0 }

        let elapsed = timestamps.last! - timestamps.first!
        guard elapsed > 0 else { return 0 }

        return Double(timestamps.count - 1) / elapsed
    }
}

File: android/src/main/kotlin/com/example/npu_pose_detection/NpuPoseDetectionPlugin.kt
package com.example.npu_pose_detection

import android.content.Context
import android.os.Build
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.*
import com.example.npu_pose_detection.models.*
import java.util.Base64

/**
 * Flutter plugin for NPU-accelerated pose detection on Android.
 *
 * Uses MediaPipe PoseLandmarker for 33-landmark detection with GPU acceleration.
 */
class NpuPoseDetectionPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {

    companion object {
        private const val TAG = "NpuPoseDetectionPlugin"
        private const val METHOD_CHANNEL = "com.example.flutter_pose_detection/methods"
        private const val EVENT_CHANNEL = "com.example.flutter_pose_detection/frames"
        private const val VIDEO_PROGRESS_CHANNEL = "com.example.flutter_pose_detection/video_progress"
    }

    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var videoProgressChannel: EventChannel
    private lateinit var context: Context
    private var mediaPipeDetector: MediaPipeNpuDetector? = null
    private var tfliteDetector: TFLitePoseDetector? = null
    private var activeDetector: PoseDetectorInterface? = null
    private var config: DetectorConfig = DetectorConfig()
    private var useNpuBackend: Boolean = false

    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    // Stream handling
    private var frameEventSink: EventChannel.EventSink? = null
    private var videoProgressEventSink: EventChannel.EventSink? = null
    private var isStreamingActive = false
    private var frameCount = 0
    private val fpsCalculator = FPSCalculator()
    private var videoProcessor: VideoProcessor? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, METHOD_CHANNEL)
        channel.setMethodCallHandler(this)

        // Setup EventChannel for camera frame streaming
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, EVENT_CHANNEL)
        eventChannel.setStreamHandler(this)

        // Setup EventChannel for video progress
        videoProgressChannel = EventChannel(flutterPluginBinding.binaryMessenger, VIDEO_PROGRESS_CHANNEL)
        videoProgressChannel.setStreamHandler(VideoProgressStreamHandler(this))
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initialize" -> handleInitialize(call, result)
            "detectPose" -> handleDetectPose(call, result)
            "detectPoseFromFile" -> handleDetectPoseFromFile(call, result)
            "processFrame" -> handleProcessFrame(call, result)
            "startCameraDetection" -> handleStartCameraDetection(result)
            "stopCameraDetection" -> handleStopCameraDetection(result)
            "analyzeVideo" -> handleAnalyzeVideo(call, result)
            "cancelVideoAnalysis" -> handleCancelVideoAnalysis(result)
            "updateConfig" -> handleUpdateConfig(call, result)
            "getDeviceCapabilities" -> handleGetDeviceCapabilities(result)
            "benchmarkDelegates" -> handleBenchmarkDelegates(call, result)
            "dispose" -> handleDispose(result)
            else -> result.notImplemented()
        }
    }

    // StreamHandler implementation
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        frameEventSink = events
    }

    override fun onCancel(arguments: Any?) {
        frameEventSink = null
        isStreamingActive = false
    }

    // MARK: - Initialize

    private fun handleInitialize(call: MethodCall, result: Result) {
        val args = call.arguments as? Map<*, *>
        val configMap = args?.get("config") as? Map<*, *>

        if (configMap == null) {
            result.success(errorResponse("invalidArguments", "Missing config"))
            return
        }

        config = DetectorConfig.fromMap(configMap)
        useNpuBackend = config.preferredAcceleration == AccelerationMode.NPU

        scope.launch {
            try {
                if (useNpuBackend) {
                    // Use TFLite + QNN for NPU acceleration (better battery efficiency)
                    Log.i(TAG, "Initializing TFLite with QNN delegate (NPU)...")
                    tfliteDetector = TFLitePoseDetector(context)
                    val mode = withContext(Dispatchers.IO) {
                        tfliteDetector?.initialize(config)
                    }

                    if (tfliteDetector?.isInitialized != true) {
                        throw IllegalStateException("TFLite detector not initialized")
                    }

                    activeDetector = tfliteDetector
                    Log.i(TAG, "✓ TFLite NPU initialized successfully, mode=$mode")
                    result.success(mapOf(
                        "success" to true,
                        "accelerationMode" to (mode?.name?.lowercase() ?: "npu"),
                        "modelVersion" to "tflite_pose_landmarks",
                        "numLandmarks" to 33
                    ))
                } else {
                    // Use MediaPipe for GPU acceleration (faster but more power)
                    Log.i(TAG, "Initializing MediaPipe PoseLandmarker...")
                    mediaPipeDetector = MediaPipeNpuDetector(context)
                    val mode = withContext(Dispatchers.IO) {
                        mediaPipeDetector?.initialize(config)
                    }

                    if (mediaPipeDetector?.isInitialized != true) {
                        throw IllegalStateException("MediaPipe PoseLandmarker not initialized")
                    }

                    activeDetector = mediaPipeDetector
                    Log.i(TAG, "✓ MediaPipe initialized successfully, mode=$mode")
                    result.success(mapOf(
                        "success" to true,
                        "accelerationMode" to (mode?.name?.lowercase() ?: "gpu"),
                        "modelVersion" to "mediapipe_pose_landmarker",
                        "numLandmarks" to 33
                    ))
                }
            } catch (e: Exception) {
                Log.e(TAG, "Initialization failed: ${e.message}", e)
                result.success(errorResponse(
                    "modelLoadFailed",
                    "Failed to initialize pose detector",
                    e.message
                ))
            }
        }
    }

    // MARK: - Detect Pose

    private fun handleDetectPose(call: MethodCall, result: Result) {
        if (!isDetectorReady()) {
            result.success(errorResponse("notInitialized", "Detector not initialized"))
            return
        }

        val args = call.arguments as? Map<*, *>
        val imageDataBase64 = args?.get("imageData") as? String

        if (imageDataBase64 == null) {
            result.success(errorResponse("invalidImageFormat", "Invalid image data"))
            return
        }

        scope.launch {
            try {
                val imageData = Base64.getDecoder().decode(imageDataBase64)
                val poseResult = withContext(Dispatchers.IO) {
                    activeDetector?.detectPose(imageData)
                }

                result.success(mapOf(
                    "success" to true,
                    "result" to poseResult?.toMap()
                ))
            } catch (e: Exception) {
                result.success(errorResponse(
                    "inferenceFailed",
                    "Pose detection failed",
                    e.message
                ))
            }
        }
    }

    // MARK: - Detect Pose From File

    private fun handleDetectPoseFromFile(call: MethodCall, result: Result) {
        if (!isDetectorReady()) {
            result.success(errorResponse("notInitialized", "Detector not initialized"))
            return
        }

        val args = call.arguments as? Map<*, *>
        val filePath = args?.get("filePath") as? String

        if (filePath == null) {
            result.success(errorResponse("invalidArguments", "Missing file path"))
            return
        }

        scope.launch {
            try {
                val poseResult = withContext(Dispatchers.IO) {
                    activeDetector?.detectPoseFromFile(filePath)
                }

                result.success(mapOf(
                    "success" to true,
                    "result" to poseResult?.toMap()
                ))
            } catch (e: Exception) {
                result.success(errorResponse(
                    "inferenceFailed",
                    "Pose detection failed",
                    e.message
                ))
            }
        }
    }

    private fun isDetectorReady(): Boolean {
        return activeDetector?.isInitialized == true
    }

    // MARK: - Update Config

    private fun handleUpdateConfig(call: MethodCall, result: Result) {
        val args = call.arguments as? Map<*, *>
        val configMap = args?.get("config") as? Map<*, *>

        if (configMap == null) {
            result.success(errorResponse("invalidArguments", "Missing config"))
            return
        }

        config = DetectorConfig.fromMap(configMap)
        mediaPipeDetector?.updateConfig(config)

        result.success(mapOf(
            "success" to true,
            "requiresReinitialize" to false
        ))
    }

    // MARK: - Device Capabilities

    private fun handleGetDeviceCapabilities(result: Result) {
        val hasNnapi = Build.VERSION.SDK_INT in 27..34
        val hasGpu = true // Most devices support GPU delegate

        result.success(mapOf(
            "success" to true,
            "capabilities" to mapOf(
                "platform" to "android",
                "osVersion" to Build.VERSION.RELEASE,
                "deviceModel" to "${Build.MANUFACTURER} ${Build.MODEL}",
                "npuInfo" to mapOf(
                    "type" to "nnapi",
                    "isAvailable" to hasNnapi,
                    "performanceTier" to if (hasNnapi) "medium" else "low"
                ),
                "hasGpuSupport" to hasGpu,
                "recommendedMode" to if (hasGpu) "gpu" else "cpu"
            )
        ))
    }

    // MARK: - Process Frame

    private fun handleProcessFrame(call: MethodCall, result: Result) {
        if (!isDetectorReady()) {
            result.success(errorResponse("notInitialized", "Detector not initialized"))
            return
        }

        val args = call.arguments as? Map<*, *>
        @Suppress("UNCHECKED_CAST")
        val planes = args?.get("planes") as? List<Map<String, Any>>
        val width = args?.get("width") as? Int
        val height = args?.get("height") as? Int
        val format = args?.get("format") as? String ?: "yuv420"
        val rotation = args?.get("rotation") as? Int ?: 0

        if (planes == null || width == null || height == null) {
            result.success(errorResponse("invalidArguments", "Missing frame data"))
            return
        }

        // Decode base64 bytes
        val decodedPlanes = planes.map { plane ->
            val bytesBase64 = plane["bytes"] as? String
            val bytes = if (bytesBase64 != null) {
                Base64.getDecoder().decode(bytesBase64)
            } else {
                ByteArray(0)
            }
            mapOf(
                "bytes" to bytes,
                "bytesPerRow" to (plane["bytesPerRow"] as? Int ?: 0),
                "bytesPerPixel" to (plane["bytesPerPixel"] as? Int ?: 1)
            )
        }

        scope.launch {
            try {
                val poseResult = withContext(Dispatchers.IO) {
                    mediaPipeDetector?.processFrame(decodedPlanes, width, height, format, rotation)
                }

                result.success(mapOf(
                    "success" to true,
                    "result" to poseResult?.toMap()
                ))
            } catch (e: Exception) {
                result.success(errorResponse(
                    "inferenceFailed",
                    "Frame processing failed",
                    e.message
                ))
            }
        }
    }

    // MARK: - Camera Detection

    private fun handleStartCameraDetection(result: Result) {
        if (!isDetectorReady()) {
            result.success(errorResponse("notInitialized", "Detector not initialized"))
            return
        }

        isStreamingActive = true
        frameCount = 0
        fpsCalculator.reset()

        result.success(mapOf("success" to true))
    }

    private fun handleStopCameraDetection(result: Result) {
        isStreamingActive = false

        // Send end event
        frameEventSink?.success(mapOf(
            "type" to "end",
            "reason" to "stopped"
        ))

        result.success(mapOf("success" to true))
    }

    // MARK: - Video Analysis

    private fun handleAnalyzeVideo(call: MethodCall, result: Result) {
        if (!isDetectorReady()) {
            result.success(errorResponse("notInitialized", "Detector not initialized"))
            return
        }

        val args = call.arguments as? Map<*, *>
        val videoPath = args?.get("videoPath") as? String
        val frameInterval = (args?.get("frameInterval") as? Int) ?: 1

        if (videoPath == null) {
            result.success(errorResponse("invalidArguments", "Missing video path"))
            return
        }

        val detector = mediaPipeDetector!! as PoseDetectorInterface

        videoProcessor = VideoProcessor(context, detector)

        scope.launch {
            try {
                val analysisResult = videoProcessor?.analyzeVideo(
                    videoPath,
                    frameInterval
                ) { progress ->
                    videoProgressEventSink?.success(progress.toMap())
                }

                if (analysisResult != null) {
                    videoProgressEventSink?.success(mapOf("type" to "complete"))
                    result.success(mapOf(
                        "success" to true,
                        "result" to analysisResult.toMap()
                    ))
                } else {
                    result.success(errorResponse("videoAnalysisFailed", "Analysis returned null"))
                }
            } catch (e: Exception) {
                videoProgressEventSink?.success(mapOf(
                    "type" to "error",
                    "error" to mapOf(
                        "code" to "videoAnalysisFailed",
                        "message" to (e.message ?: "Unknown error")
                    )
                ))
                result.success(errorResponse(
                    "videoAnalysisFailed",
                    "Video analysis failed",
                    e.message
                ))
            }
        }
    }

    private fun handleCancelVideoAnalysis(result: Result) {
        videoProcessor?.cancel()
        videoProgressEventSink?.success(mapOf("type" to "cancelled"))
        result.success(mapOf("success" to true))
    }

    // MARK: - Benchmark Delegates

    private fun handleBenchmarkDelegates(call: MethodCall, result: Result) {
        val args = call.arguments as? Map<*, *>
        val iterations = (args?.get("iterations") as? Int) ?: 10

        Log.i(TAG, "Starting delegate benchmark (iterations=$iterations)...")

        scope.launch {
            try {
                val tfliteDetector = TFLitePoseDetector(context)
                val results = withContext(Dispatchers.IO) {
                    tfliteDetector.benchmarkAll(iterations)
                }
                tfliteDetector.dispose()

                val benchmarkResults = results.map { (delegate, benchResult) ->
                    delegate to mapOf(
                        "success" to benchResult.success,
                        "avgInferenceTimeMs" to benchResult.avgInferenceTimeMs,
                        "minInferenceTimeMs" to benchResult.minInferenceTimeMs,
                        "maxInferenceTimeMs" to benchResult.maxInferenceTimeMs,
                        "errorMessage" to benchResult.errorMessage
                    )
                }.toMap()

                result.success(mapOf(
                    "success" to true,
                    "results" to benchmarkResults
                ))
            } catch (e: Exception) {
                Log.e(TAG, "Benchmark failed: ${e.message}", e)
                result.success(errorResponse(
                    "benchmarkFailed",
                    "Benchmark failed",
                    e.message
                ))
            }
        }
    }

    // MARK: - Dispose

    private fun handleDispose(result: Result) {
        isStreamingActive = false
        videoProcessor?.cancel()
        videoProcessor = null
        activeDetector = null
        mediaPipeDetector?.dispose()
        mediaPipeDetector = null
        tfliteDetector?.dispose()
        tfliteDetector = null
        useNpuBackend = false
        result.success(mapOf("success" to true))
    }

    // Public method for VideoProgressStreamHandler
    fun setVideoProgressEventSink(sink: EventChannel.EventSink?) {
        videoProgressEventSink = sink
    }

    // MARK: - Helpers

    private fun errorResponse(
        code: String,
        message: String,
        platformMessage: String? = null
    ): Map<String, Any> {
        val error = mutableMapOf<String, Any>(
            "code" to code,
            "message" to message
        )
        platformMessage?.let { error["platformMessage"] = it }

        return mapOf(
            "success" to false,
            "error" to error
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        videoProgressChannel.setStreamHandler(null)
        scope.cancel()
        activeDetector = null
        mediaPipeDetector?.dispose()
        mediaPipeDetector = null
        tfliteDetector?.dispose()
        tfliteDetector = null
    }
}

/**
 * Stream handler for video progress events.
 */
class VideoProgressStreamHandler(
    private val plugin: NpuPoseDetectionPlugin
) : EventChannel.StreamHandler {

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        plugin.setVideoProgressEventSink(events)
    }

    override fun onCancel(arguments: Any?) {
        plugin.setVideoProgressEventSink(null)
    }
}

/**
 * FPS calculator for streaming.
 */
class FPSCalculator {
    private val timestamps = mutableListOf<Long>()
    private val windowSize = 30

    fun reset() {
        timestamps.clear()
    }

    fun update(): Double {
        val now = System.currentTimeMillis()
        timestamps.add(now)

        // Keep only recent timestamps
        while (timestamps.size > windowSize) {
            timestamps.removeAt(0)
        }

        // Calculate FPS from time window
        if (timestamps.size < 2) return 0.0

        val elapsed = timestamps.last() - timestamps.first()
        if (elapsed <= 0) return 0.0

        return (timestamps.size - 1) * 1000.0 / elapsed
    }
}
